### PR TITLE
Issue #3091 - Add etcd dependency TLS support

### LINKED ISSUE #3091: Add TLS support for etcd in ring
I recently worked on some minor updates that add TLS support to **etcd** . Would like to create a PR to submit these updates for review.

### PR DESCRIPTION
**Pull request by maunavarro**
**What this PR does**:
Adds TLS support for **etcd** dependency

**Which issue(s) this PR fixes**:
Fixes #3091 

**Checklist**
- [x] Tests updated
- [x] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: pkg/ring/kv/etcd/etcd.go
Comment (pracucci) [inline pkg/ring/kv/etcd/etcd.go:51]: I would merge `SetTLS()` logic into `GetTLS()`. No need to store TLS config in `cfg.TLS`. You can just have `GetTLS()` generating it on the fly and then returning `*transport.TLSInfo`.

**Review by pracucci — COMMENTED** [review]
Thanks for working on this! Much appreciated! I left few comments, but overall looks good.

A couple of more things:
1. Should we also allow to configure `InsecureSkipVerify` for people using self-signed certs?
2. How can we test it?

File: pkg/ring/kv/etcd/etcd.go
Comment (pstibrany) [reply pkg/ring/kv/etcd/etcd.go:52]: Let's please use an explicit flag for enabling TLS. It should be possible to enable TLS with any of these settings (Cert+Key are for client authentication, while TrustedCAFile being empty means to use trusted roots from OS).

File: docs/configuration/arguments.md
Comment (pracucci) [inline docs/configuration/arguments.md:187]: Would you mind updating these descriptions with the new descriptions we put in the code, please?

File: pkg/ring/kv/etcd/etcd.go
Comment (pracucci) [inline pkg/ring/kv/etcd/etcd.go:68]: We can simplify this.

```suggestion
	tlsConfig.InsecureSkipVerify = cfg.InsecureSkipVerify
```

File: pkg/ring/kv/etcd/etcd.go
Comment (pracucci) [inline pkg/ring/kv/etcd/etcd.go:76]: In case an error occurs, we should give an hint to the user to which part of the initialisation fail. The error we may get from `GetTLS()` could be very generic.

```suggestion
		return nil, errors.Wrapf(err, "unable to initialise TLS configuration for etcd")
```

File: pkg/ring/kv/etcd/etcd.go
Comment (pracucci) [inline pkg/ring/kv/etcd/etcd.go:43]:
```suggestion
```

File: pkg/ring/kv/etcd/etcd.go
Comment (pracucci) [inline pkg/ring/kv/etcd/etcd.go:27]: Please add the CLI flag for this too

File: pkg/ring/kv/etcd/etcd.go
Comment (pracucci) [inline pkg/ring/kv/etcd/etcd.go:24]: I would add the `tls_` prefix to all such config options and `tls-` to CLI flags.

File: pkg/ring/kv/etcd/etcd.go
Comment (pracucci) [inline pkg/ring/kv/etcd/etcd.go:51]: @pstibrany feedback was to add another option `tls_enabled` (defaults to false) to explicitely turn it on, because on the client side (here) you may want to turn on TLS without specifying any cert/key/CA.

**Review by pracucci — COMMENTED** [review]
Could you also add a `CHANGELOG` entry (`[ENHANCEMENT]`), please?

File: pkg/ring/kv/etcd/etcd.go
Comment (maunavarro) [reply pkg/ring/kv/etcd/etcd.go:51]: Ah, makes sense. Working on this. Thanks.

File: CHANGELOG.md
Comment (pracucci) [inline CHANGELOG.md:11]: This should be here. Please rebase `master` instead of merging it.

File: CHANGELOG.md
Comment (pracucci) [inline CHANGELOG.md:31]:
```suggestion
* [ENHANCEMENT] Add TLS support to etcd client (used by the hash ring). #3091
```

File: docs/configuration/arguments.md
Comment (pracucci) [inline docs/configuration/arguments.md:195]: Uhm, server or client? Isn't to skip the verification of the provided `-etcd.tls-cert-file`?

File: pkg/ring/kv/etcd/etcd.go
Comment (pracucci) [inline pkg/ring/kv/etcd/etcd.go:28]: Sorry for being annoying but I would like to keep consistency with existing TLS config (eg. see https://cortexmetrics.io/docs/configuration/configuration-file/#ruler_config).

```suggestion
	CertFile           string        `yaml:"tls_cert_path"`
	KeyFile            string        `yaml:"tls_key_path"`
	TrustedCAFile      string        `yaml:"tls_ca_path"`
```

File: pkg/ring/kv/etcd/etcd.go
Comment (pracucci) [inline pkg/ring/kv/etcd/etcd.go:48]: Sorry for being annoying but I would like to keep consistency with existing TLS config (eg. see https://cortexmetrics.io/docs/configuration/configuration-file/#ruler_config).

```suggestion
	f.StringVar(&cfg.CertFile, prefix+"etcd.tls-cert-path", "", "The TLS certificate file path")
	f.StringVar(&cfg.KeyFile, prefix+"etcd.tls-key-path", "", "The TLS private key file path")
	f.StringVar(&cfg.TrustedCAFile, prefix+"etcd.tls-ca-path", "", "The trusted CA file path")
```

File: pkg/ring/kv/etcd/etcd.go
Comment (pracucci) [inline pkg/ring/kv/etcd/etcd.go:54]: Here you should just check for `if !cfg.EnableTLS`.

File: pkg/ring/kv/etcd/etcd.go
Comment (pracucci) [inline pkg/ring/kv/etcd/etcd.go:62]: No need for the `if` because the empty string is also the zero value of the string in the `transport.TLSInfo` structure.

```suggestion
		tlsInfo.TrustedCAFile = cfg.TrustedCAFile
```

File: pkg/ring/kv/etcd/etcd.go
Comment (pracucci) [inline pkg/ring/kv/etcd/etcd.go:68]: Move this up, where `&transport.TLSInfo`  is initialized.

File: docs/configuration/arguments.md
Comment (pracucci) [inline docs/configuration/arguments.md:192]:
```suggestion
- `etcd.tls-cert-path`
   The TLS certificate file path.
- `etcd.tls-key-path`
   The TLS private key file path.
- `etcd.tls-ca-path`
```

**Review by pracucci — APPROVED** [review]
Thanks for working on this and patiently address my feedback! LGTM with a couple of things:
1. Please rebase `master` (not merge it)
2. `make doc` again
3. Left a suggestion to fix `docs/configuration/arguments.md`

**Comment by maunavarro** [comment]
@pracucci thank you for PR review and the comments. I went ahead and did a rebase, ran `make doc` and added your suggestions to fix `docs/configuration/arguments.md`. Let me know of any additional comments / updates. Thanks again.

File: docs/configuration/arguments.md
Comment (pstibrany) [reply docs/configuration/arguments.md]: This flag skips validation of certificate presented by server, so now the description is misleading. Please revert it back to "Skip validating server certificate."

File: CHANGELOG.md
Comment (pstibrany) [inline CHANGELOG.md:31]: We normally use PR numbers in the changelog, so this should be

```suggestion
* [ENHANCEMENT] Add TLS support to etcd client. #3102
```

(Also etcd client is used at different places, eg. HA Tracker, so I've removed "used by the hash ring" part).

File: pkg/ring/kv/etcd/etcd.go
Comment (pstibrany) [inline pkg/ring/kv/etcd/etcd.go:49]: Nit: please add periods at the end of description sentences for consistency.

File: pkg/ring/kv/etcd/etcd.go
Comment (pstibrany) [inline pkg/ring/kv/etcd/etcd.go:67]:
```suggestion
	return tlsInfo.ClientConfig()
```

**Review by pstibrany — COMMENTED** [review]
Thank you! Looks good overall. I've left some tiny comments, once fixed, we can merge.

**Comment by maunavarro** [comment]
@pstibrany thanks for the review. I went ahead and addressed your comments. Let me know.

File: pkg/ring/kv/etcd/etcd.go
Comment (pstibrany) [inline pkg/ring/kv/etcd/etcd.go:49]:
```suggestion
	f.BoolVar(&cfg.InsecureSkipVerify, prefix+"etcd.tls-insecure-skip-verify", false, "Skip validating server certificate.")
```

File: docs/configuration/arguments.md
Comment (pstibrany) [inline docs/configuration/arguments.md:195]:
```suggestion
   Skip validating server certificate.
```

**Review by pstibrany — COMMENTED** [review]
Thank you!

Description for `etcd.tls-insecure-skip-verify` talks about client certificate, which is incorrect -- it skips validation of certificate received from the server. Let's fix that.

**Comment by maunavarro** [comment]
@pstibrany had missed the `etcd.tls-insecure-skip-verify` update; went ahead and fixed it. Let me know if you see anything else. Thanks!

**Review by pstibrany — APPROVED** [review]
Thank you very much for your work and patience!

**Pull request merged**
